src/main/scala/org/neo4j/spark/Executor.scala

```diff
@@ -48,75 +37,39 @@ object Executor {
     execute(Neo4jConfig(sc.getConf), query, parameters)
   }
 
-  private def rows(result: StatementResult) = {
+  private def rows(result: Iterator[_]) = {
     var i = 0
     while (result.hasNext) i = i + 1
     i
   }
 
   def execute(config: Neo4jConfig, query: String, parameters: Map[String, Any], write: Boolean = false): CypherResult = {
-
-    def close(driver: Driver, session: Session) = {
-      try {
-        if (session.isOpen) {
-          session.close()
-        }
-        driver.close()
-      } catch {
-        case _ => // ignore
-      }
+    val result = new Neo4jSessionAwareIterator(config, query, toJava(parameters), write)
+    if (!result.hasNext) {
+      return EMPTY_RESULT
     }
-
-    val driver: Driver = config.driver()
-    val session = driver.session()
-
-    try {
-      val runner = new TransactionWork[CypherResult]() {
-        override def execute(tx: Transaction): CypherResult = {
-          val result: StatementResult = tx.run(query, toJava(parameters))
-          if (!result.hasNext) {
-            result.consume()
-            session.close()
-            driver.close()
-            return new CypherResult(new StructType(), Iterator.empty)
-          }
-          val peek = result.peek()
-          val keyCount = peek.size()
-          if (keyCount == 0) {
-            val res: CypherResult = new CypherResult(new StructType(), Array.fill[Array[Any]](rows(result))(EMPTY).toIterator)
-            result.consume()
-            close(driver, session)
-            return res
-          }
-          val keys = peek.keys().asScala
-          val fields = keys.map(k => (k, peek.get(k).`type`())).map(keyType => CypherTypes.field(keyType))
-          val schema = StructType(fields)
-
-          val it = result.asScala.map(record => {
-            val row = new Array[Any](keyCount)
-            var i = 0
-            while (i < keyCount) {
-              val value = convert(record.get(i).asObject())
```

Comment:
pleas re-add this bit

Reply:
I'm sorry something went wrong when I rebased the code.
I'll re-add this and others part